core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -284,6 +283,11 @@ private void decPendingAndFinishIfNeeded() {
 
     private void finish() {
         if (finished.compareAndSet(false, true)) {
+            try {
+                primary.updateGlobalCheckpoint();
```

Comment:
>  We should in the future also consider the case where the background sync fails

I think this merits a good follow up discussion. This PR was started with the think of the global checkpoint as an optimization - i.e., it's not a big deal if it lags, even indefinitely. This is similar (albeit different) to the decision to not fail shards/mark them as stale on a failure to sync it. However, since we plan to build infrastructure on this information, I think we should re-evaluate it and get to a place where you can reasonably expect the global checkpoint to eventually be updated.